File: componenthub_mcp/providers/mouser.py
"""Mouser connector — Search API v1 (API key).

Capabilities: search, details, pricing, availability, datasheet.
Enable with MOUSER_API_KEY (mouser.com/api-hub).
"""

import re

import httpx

from .. import config
from ..models import (
    CadAsset,
    Capability,
    ComponentDetails,
    ComponentResult,
    Offer,
    PriceBreak,
    SearchQuery,
)
from .base import Provider, ProviderError

_API = "https://api.mouser.com/api/v1"


def _parse_price(text: str | None) -> float | None:
    if not text:
        return None
    cleaned = re.sub(r"[^\d.,]", "", text).replace(",", ".")
    try:
        return float(cleaned)
    except ValueError:
        return None


class MouserProvider(Provider):
    name = "mouser"
    display_name = "Mouser Electronics"
    capabilities = frozenset(
        {
            Capability.SEARCH,
            Capability.DETAILS,
            Capability.PRICING,
            Capability.AVAILABILITY,
            Capability.DATASHEET,
        }
    )

    def is_configured(self) -> bool:
        return bool(config.mouser_api_key())

    def missing_config(self) -> str | None:
        if self.is_configured():
            return None
        return "Set MOUSER_API_KEY (mouser.com/api-hub)"

    def _map_part(self, p: dict) -> ComponentResult:
        breaks = []
        for b in p.get("PriceBreaks") or []:
            price = _parse_price(b.get("Price"))
            if price is not None:
                breaks.append(
                    PriceBreak(
                        quantity=b.get("Quantity", 1),
                        unit_price=price,
                        currency=b.get("Currency", "USD"),
                    )
                )
        stock_match = re.search(r"\d+", p.get("Availability") or "")
        offer = Offer(
            provider=self.name,
            part_id=p.get("MouserPartNumber") or p.get("ManufacturerPartNumber", ""),
            product_url=p.get("ProductDetailUrl"),
            stock=int(stock_match.group()) if stock_match else None,
            price_breaks=breaks,
        )
        return ComponentResult(
            mpn=p.get("ManufacturerPartNumber", ""),
            manufacturer=p.get("Manufacturer"),
            description=p.get("Description"),
            category=p.get("Category"),
            datasheet_url=p.get("DataSheetUrl") or None,
            image_url=p.get("ImagePath") or None,
            offers=[offer],
        )

    async def _keyword_search(self, keyword: str, records: int) -> list[dict]:
        async with httpx.AsyncClient(timeout=20) as client:
            resp = await client.post(
                f"{_API}/search/keyword",
                params={"apiKey": config.mouser_api_key()},
                json={
                    "SearchByKeywordRequest": {
                        "keyword": keyword,
                        "records": records,
                        "startingRecord": 0,
                    }
                },
            )
            if resp.status_code != 200:
                raise ProviderError(f"mouser: search failed ({resp.status_code}): {resp.text[:200]}")
            data = resp.json()
            errors = data.get("Errors") or []
            if errors:
                raise ProviderError(f"mouser: {errors[0].get('Message', 'API error')}")
            return (data.get("SearchResults") or {}).get("Parts") or []

    async def search(self, query: SearchQuery) -> list[ComponentResult]:
        parts = await self._keyword_search(query.keyword, query.max_results)
        results = [self._map_part(p) for p in parts]
        if query.manufacturer:
            results = [
                r for r in results
                if r.manufacturer and query.manufacturer.lower() in r.manufacturer.lower()
            ]
        if query.in_stock_only:
            results = [r for r in results if any((o.stock or 0) > 0 for o in r.offers)]
        return results

    async def fetch_details(self, part_id: str) -> ComponentDetails:
        parts = await self._keyword_search(part_id, 1)
        if not parts:
            raise ProviderError(f"mouser: part {part_id!r} not found")
        p = parts[0]
        base = self._map_part(p)
        specs = {
            (attr.get("AttributeName") or ""): (attr.get("AttributeValue") or "")
            for attr in p.get("ProductAttributes") or []
        }
        return ComponentDetails(
            **base.model_dump(),
            specifications=specs,
            lifecycle_status=p.get("LifecycleStatus") or None,
            cad_assets=[],
        )

    async def fetch_pricing(self, part_id: str) -> Offer:
        details = await self.fetch_details(part_id)
        if not details.offers:
            raise ProviderError("mouser: no pricing returned")
        return details.offers[0]

    async def fetch_datasheet(self, part_id: str) -> str | None:
        return (await self.fetch_details(part_id)).datasheet_url

    async def fetch_models(self, part_id: str) -> list[CadAsset]:
        raise ProviderError("mouser: CAD models not offered; use snapmagic")

File: lib/domain/providers/mouser.ts
/**
 * Mouser connector — Search API v1 (API key).
 * Ports componenthub_mcp/providers/mouser.py exactly.
 *
 * Capabilities: search, details, pricing, availability, datasheet.
 * Enable with MOUSER_API_KEY (mouser.com/api-hub).
 */
import { mouserApiKey } from "../config";
import type {
  CadAsset,
  ComponentDetails,
  ComponentResult,
  Offer,
  SearchQuery,
} from "../models";
import { Capability } from "../models";
import { Provider, ProviderError } from "./base";

const API = "https://api.mouser.com/api/v1";

function parsePrice(text: string | null | undefined): number | null {
  if (!text) return null;
  const cleaned = text.replace(/[^\d.,]/g, "").replace(",", ".");
  const n = parseFloat(cleaned);
  return Number.isNaN(n) ? null : n;
}

export class MouserProvider extends Provider {
  readonly name = "mouser";
  readonly displayName = "Mouser Electronics";
  readonly capabilities = new Set([
    Capability.SEARCH,
    Capability.DETAILS,
    Capability.PRICING,
    Capability.AVAILABILITY,
    Capability.DATASHEET,
  ]);

  isConfigured(): boolean {
    return Boolean(mouserApiKey());
  }

  missingConfig(): string | null {
    if (this.isConfigured()) return null;
    return "Set MOUSER_API_KEY (mouser.com/api-hub).";
  }

  private _mapPart(p: Record<string, unknown>): ComponentResult {
    const priceBreaks = (p.PriceBreaks as Record<string, unknown>[]) ?? [];
    const breaks: { quantity: number; unit_price: number; currency: string }[] =
      [];
    for (const b of priceBreaks) {
      const price = parsePrice(b.Price as string);
      if (price !== null) {
        breaks.push({
          quantity: (b.Quantity as number) ?? 1,
          unit_price: price,
          currency: (b.Currency as string) ?? "USD",
        });
      }
    }
    const availability = (p.Availability as string) ?? "";
    const stockMatch = availability.match(/\d+/);
    const offer: Offer = {
      provider: this.name,
      part_id:
        (p.MouserPartNumber as string) ??
        (p.ManufacturerPartNumber as string) ??
        "",
      product_url: (p.ProductDetailUrl as string) ?? null,
      stock: stockMatch ? parseInt(stockMatch[0], 10) : null,
      price_breaks: breaks,
    };
    return {
      mpn: (p.ManufacturerPartNumber as string) ?? "",
      manufacturer: (p.Manufacturer as string) ?? null,
      description: (p.Description as string) ?? null,
      category: (p.Category as string) ?? null,
      datasheet_url: (p.DataSheetUrl as string) || null,
      image_url: (p.ImagePath as string) || null,
      offers: [offer],
    };
  }

  private async _keywordSearch(
    keyword: string,
    records: number,
  ): Promise<Record<string, unknown>[]> {
    const apiKey = mouserApiKey();
    const resp = await fetch(
      `${API}/search/keyword?apiKey=${encodeURIComponent(apiKey ?? "")}`,
      {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({
          SearchByKeywordRequest: {
            keyword,
            records,
            startingRecord: 0,
          },
        }),
      },
    );
    if (!resp.ok) {
      const text = await resp.text();
      throw new ProviderError(
        `mouser: search failed (${resp.status}): ${text.slice(0, 200)}`,
      );
    }
    const data = (await resp.json()) as {
      Errors?: { Message?: string }[];
      SearchResults?: { Parts?: Record<string, unknown>[] };
    };
    const errors = data.Errors ?? [];
    if (errors.length) {
      throw new ProviderError(
        `mouser: ${errors[0].Message ?? "API error"}`,
      );
    }
    return data.SearchResults?.Parts ?? [];
  }

  async search(query: SearchQuery): Promise<ComponentResult[]> {
    const parts = await this._keywordSearch(query.keyword, query.max_results);
    let results = parts.map((p) => this._mapPart(p));
    if (query.manufacturer) {
      const needle = query.manufacturer.toLowerCase();
      results = results.filter(
        (r) => r.manufacturer?.toLowerCase().includes(needle),
      );
    }
    if (query.in_stock_only) {
      results = results.filter((r) =>
        r.offers.some((o) => (o.stock ?? 0) > 0),
      );
    }
    return results;
  }

  async fetchDetails(partId: string): Promise<ComponentDetails> {
    const parts = await this._keywordSearch(partId, 1);
    if (!parts.length) {
      throw new ProviderError(`mouser: part ${JSON.stringify(partId)} not found`);
    }
    const p = parts[0];
    const base = this._mapPart(p);
    const attrs = (p.ProductAttributes as Record<string, unknown>[]) ?? [];
    const specs: Record<string, string> = {};
    for (const attr of attrs) {
      const key = (attr.AttributeName as string) ?? "";
      const val = (attr.AttributeValue as string) ?? "";
      specs[key] = val;
    }
    return {
      ...base,
      specifications: specs,
      lifecycle_status: (p.LifecycleStatus as string) ?? null,
      cad_assets: [],
    };
  }

  async fetchPricing(partId: string): Promise<Offer> {
    const details = await this.fetchDetails(partId);
    if (!details.offers.length) {
      throw new ProviderError("mouser: no pricing returned");
    }
    return details.offers[0];
  }

  async fetchDatasheet(partId: string): Promise<string | null> {
    return (await this.fetchDetails(partId)).datasheet_url ?? null;
  }

  async fetchModels(_partId: string): Promise<CadAsset[]> {
    throw new ProviderError("mouser: CAD models not offered; use snapmagic");
  }
}
